proof-generator/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+use anyhow::anyhow;
+use coins_bip32::path::DerivationPath;
+use ethers::{
+    prelude::Signer,
+    providers::{Http, Middleware, Provider},
+    signers::{HDPath, Ledger},
+    types::{Signature, H256},
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    convert::TryFrom,
+    fmt::{Debug, Display},
+    fs,
+    io::Write,
+    path::PathBuf,
+    process::{Command, Stdio},
+    str,
+};
+
+/// The options allowed to be provided to the CLI
+#[derive(Debug, Clone)]
+pub struct Options {
+    /// GPG key to sign the proof.
+    pub gpg_key: String,
+    /// Output path of created proof
+    pub output: PathBuf,
+    /// RPC url
+    pub rpc_url: String,
+    /// Account derivation path when using a Ledger hardware wallet.
+    pub ledger_hdpath: Option<DerivationPath>,
+    /// Keystore file containing encrypted private key (default: none).
+    pub keystore: Option<PathBuf>,
+}
+
+/// The proof to be constructed and signed
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Proof {
+    /// GPG signature of predefined proof
+    gpg_sig: String,
+    /// ETH signature of predefined proof
+    eth_sig: Signature,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// No wallet specified.
+    #[error("no wallet specified")]
+    NoWallet,
+    /// Not able to retrieve block .
+    #[error("not able to retrieve block")]
+    NoBlock,
+    /// Not able to retrieve block hash .
+    #[error("not able to retrieve block hash")]
+    NoBlockHash,
+    /// ETH signature failed
+    #[error("eth signature failed")]
+    ETHSigFailed,
+    /// GPG signature failed
+    #[error("{0}")]
+    GPGSigFailed(String),
+}
+
+/// Sign a message with a GPG private key using the GPG CLI
+fn gpg_sign(key: &str, proof: &str) -> anyhow::Result<String> {
```

Comment:
The term "proof" seems to be overloaded, here it's simply a *message*, the proof is what you have above under `Proof` plus the keys.